multitimeseries_table.go

```
@@ -25,44 +44,82 @@ func (o *multiTimeSeriesT) Set(v interface{}) Op {
 	bucket := o.bucket(timeStamp.Unix())
-	return o.Where(Eq(o.indexField, v), Eq(bucketFieldName, bucket), Eq(o.timeField, timeStamp), Eq(o.idField, id)).ReadOne(pointer)
-
+	relations := fRelations(o.indexes(v), Eq(bucketFieldName, bucket), Eq(o.timeField, timeStamp), Eq(o.idField, id))
+	return o.Where(relations...).ReadOne(pointer)
 }
 
 func (o *multiTimeSeriesT) List(v interface{}, startTime time.Time, endTime time.Time, pointerToASlice interface{}) Op {
 	buckets := []interface{}{}
 	start := o.bucket(startTime.Unix())
-	for i := start; ; i += int64(o.bucketSize/time.Second) * 1000 {
-		if i >= endTime.Unix()*1000 {
-			break
-		}
+	end := o.bucket(endTime.Unix())
+	for i := start; i <= end; i = o.bucketer.Next(i) { // nearly but not quite an iterator
 		buckets = append(buckets, i)
 	}
-	return o.Where(Eq(o.indexField, v), In(bucketFieldName, buckets...), GTE(o.timeField, startTime), LTE(o.timeField, endTime)).Read(pointerToASlice)
+	relations := fRelations(o.indexes(v), In(bucketFieldName, buckets...), GTE(o.timeField, startTime), LTE(o.timeField, endTime))
+	return o.Where(relations...).Read(pointerToASlice)
 }
 
 func (o *multiTimeSeriesT) WithOptions(opt Options) MultiTimeSeriesTable {
 	return &multiTimeSeriesT{
-		Table:      o.Table.WithOptions(opt),
-		indexField: o.indexField,
-		timeField:  o.timeField,
-		idField:    o.idField,
-		bucketSize: o.bucketSize,
+		Table:       o.Table.WithOptions(opt),
+		indexFields: o.indexFields,
+		timeField:   o.timeField,
+		idField:     o.idField,
+		bucketer:    o.bucketer,
 	}
 }
+
+func (o *multiTimeSeriesT) indexes(iv interface{}) []Relation {
+	var indexes []Relation
+	v := o.indexesAsMap(iv)
+	ni := 0
+	for _, i := range o.indexFields {
+		if vv, ok := v[i]; ok {
+			indexes = append(indexes, Eq(i, vv))
+			ni++
+		}
+	}
+	if ni != len(o.indexFields) {
+		panic("Indexes incomplete: " + strings.Join(o.indexFields, ","))
```

Should this ever happen? If so maybe it would be better to return an error and return this to the user somehow.